src/storage/index/hnsw_index.cpp

```diff
@@ -0,0 +1,364 @@
+    lowerLayer->insert(transaction, offset, lowerEntryPoint, lowerVisited);
+    const auto rand = randomEngine.nextRandomInteger(INSERT_TO_UPPER_LAYER_RAND_UPPER_BOUND);
+    if (rand <= INSERT_TO_UPPER_LAYER_RAND_UPPER_BOUND * config.upperLayerNodePct) {
+        upperLayer->insert(transaction, offset, upperLayer->getEntryPoint(), upperVisited);
+    }
+}
+
+// NOLINTNEXTLINE(readability-make-member-function-const): Semantically non-const function.
+void InMemHNSWIndex::shrink(transaction::Transaction* transaction) {
+    upperLayer->shrink(transaction);
+    lowerLayer->shrink(transaction);
+}
+
+// NOLINTNEXTLINE(readability-make-member-function-const): Semantically non-const function.
+void InMemHNSWIndex::finalize(MemoryManager& mm,
+    const HNSWIndexPartitionerSharedState& partitionerSharedState) {
+    upperLayer->finalize(mm, *partitionerSharedState.upperPartitionerSharedState);
+    lowerLayer->finalize(mm, *partitionerSharedState.lowerPartitionerSharedState);
+}
+
+OnDiskHNSWIndex::OnDiskHNSWIndex(main::ClientContext* context,
+    catalog::NodeTableCatalogEntry* nodeTableEntry, common::column_id_t columnID,
+    catalog::RelTableCatalogEntry* upperRelTableEntry,
+    catalog::RelTableCatalogEntry* lowerRelTableEntry, HNSWIndexConfig config)
+    : HNSWIndex{std::move(config)}, nodeTableID{nodeTableEntry->getTableID()},
+      upperRelTableID{upperRelTableEntry->getTableID()},
+      lowerRelTableID{lowerRelTableEntry->getTableID()} {
+    auto& nodeTable =
+        context->getStorageManager()->getTable(nodeTableEntry->getTableID())->cast<NodeTable>();
+    const auto& indexColumnType = nodeTable.getColumn(columnID).getDataType();
+    KU_ASSERT(indexColumnType.getLogicalTypeID() == common::LogicalTypeID::ARRAY);
+    const auto extraTypeInfo =
+        indexColumnType.getExtraTypeInfo()->constPtrCast<common::ArrayTypeInfo>();
+    EmbeddingTypeInfo indexColumnTypeInfo{extraTypeInfo->getChildType().copy(),
+        extraTypeInfo->getNumElements()};
+    embeddings =
+        std::make_unique<OnDiskEmbeddings>(std::move(indexColumnTypeInfo), nodeTable, columnID);
+    graph::GraphEntry lowerGraphEntry{{nodeTableEntry}, {lowerRelTableEntry}};
+    lowerGraph = std::make_unique<graph::OnDiskGraph>(context, std::move(lowerGraphEntry));
+    graph::GraphEntry upperGraphEntry{{nodeTableEntry}, {upperRelTableEntry}};
+    upperGraph = std::make_unique<graph::OnDiskGraph>(context, std::move(upperGraphEntry));
+}
+
+std::vector<NodeWithDistance> OnDiskHNSWIndex::search(transaction::Transaction* transaction,
+    const std::vector<float>& queryVector, common::length_t k, const QueryHNSWConfig& config,
+    VisitedState& visited, NodeTableScanState& embeddingScanState) const {
+    auto entryPoint = searchNNInUpperLayer(transaction, &queryVector[0], embeddingScanState);
+    if (entryPoint == common::INVALID_OFFSET) {
+        entryPoint = lowerEntryPoint;
```

Comment:
Now that I understand why you have a lowerEntryPoint, maybe rename it to defaultLowerEntryPoint because under normal conditions this just should not happen.

Reply:
Done.